### PR TITLE
bpo-41798: Allocate the _curses._C_API on the heap memory.

### PR DESCRIPTION
**Pull request by shihai1991**
https://bugs.python.org/issue41798

### CODE REVIEW

File: Modules/_cursesmodule.c [hunk 1]
```diff
@@ -4732,9 +4732,28 @@ PyInit__curses(void)
         return NULL;
     ModDict = d; /* For PyCurses_InitScr to use later */
 
+    void **PyCurses_API = PyMem_Calloc(PyCurses_API_pointers, sizeof(void *));
```
Comment (vstinner) [inline Modules/_cursesmodule.c:4736]: This API is weird. Would you be interested to add a structure of 4 pointers? Something like PyDateTime_CAPI.

The 4 pointers are used by:

```
#define PyCursesWindow_Type (*(PyTypeObject *) PyCurses_API[0])
#define PyCursesSetupTermCalled  {if (! ((int (*)(void))PyCurses_API[1]) () ) return NULL;}
#define PyCursesInitialised      {if (! ((int (*)(void))PyCurses_API[2]) () ) return NULL;}
#define PyCursesInitialisedColor {if (! ((int (*)(void))PyCurses_API[3]) () ) return NULL;}
```

Maybe you can name the 4 members as:

```
PyTypeObject *Window_Type
int SetupTermCalled(void)
int Initialised(void)
int InitialisedColor(void)
```

Or I can merge this PR and then you work on a new PR to add a structure. As you want, tell me what do you prefer ;-)

File: Modules/_cursesmodule.c [hunk 2]
```diff
@@ -4732,9 +4732,28 @@ PyInit__curses(void)
...
+    if (PyCurses_API == NULL) {
+        PyErr_NoMemory();
+        return NULL;
+    }
+    /* Initialize the C API pointer array */
+    PyCurses_API[0] = (void *)&PyCursesWindow_Type; // borrowed ref
```
Comment (vstinner) [inline Modules/_cursesmodule.c]: Would it be possible to hold a strong reference and call Py_DECREF() in curses_destructor?

File: Modules/_cursesmodule.c (hunk 1 above)
Comment (shihai1991) [reply Modules/_cursesmodule.c:4736]: > Or I can merge this PR and then you work on a new PR to add a structure. As you want, tell me what do you prefer ;-)

Oh, I perfer to use another PR to add this structure in this weekend :)

File: Modules/_cursesmodule.c (hunk 2 above)
Comment (shihai1991) [reply Modules/_cursesmodule.c]: Of course, updated.

File: Modules/_cursesmodule.c [hunk 3]
```diff
@@ -4732,9 +4733,28 @@ PyInit__curses(void)
...
+    PyCurses_API[2] = (void *)func_PyCursesInitialised;
+    PyCurses_API[3] = (void *)func_PyCursesInitialisedColor;
+
     /* Add a capsule for the C API */
-    c_api_object = PyCapsule_New(PyCurses_API, PyCurses_CAPSULE_NAME, NULL);
-    PyDict_SetItemString(d, "_C_API", c_api_object);
+    c_api_object = PyCapsule_New(PyCurses_API, PyCurses_CAPSULE_NAME,
+                                 curses_destructor);
+    if (c_api_object == NULL) {
+        PyMem_Free(PyCurses_API);
```
Comment (vstinner) [inline Modules/_cursesmodule.c:4752]: Hum, there is no a refleak here. Does it work to call curses_destructor() here?

Comment (shihai1991) [reply Modules/_cursesmodule.c:4752]: Oh, thanks, updated.
Weird. `./python Lib/test/regrtest.py -ucurses test_curses -vv -R 3:3` can not catch this refleak.

Comment (shihai1991) [reply Modules/_cursesmodule.c:4752]: > Does it work to call curses_destructor() here?

No, it's can not work because capsule object haven't created

Comment (vstinner) [reply Modules/_cursesmodule.c:4752]: >  Weird. ./python Lib/test/regrtest.py -ucurses test_curses -vv -R 3:3 can not catch this refleak.

test_curses doesn't test memory allocation failures. See _testcapi.set_nomemory() for that.

Comment (vstinner) [reply Modules/_cursesmodule.c:4752]: Ah, and _curses doesn't use multiphase init API yet.

Comment (shihai1991) [reply Modules/_cursesmodule.c:4752]: Got it.

**Pull request merged**